[build-script] Determine HOST_CC in build-script

modocache opened the pull request:
[SR-237](https://bugs.swift.org/browse/SR-237) calls for `build-script` and `build-script-impl` to be merged. This commit takes another step towards that goal by moving the logic that finds the path to the `clang` and `clang++` executables up into Python-land.

Rather than simply moving all of the logic into `utils/build-script`, this commit moves relevant functions into a new Python module, named `buildcli`. This has several benefits:

- The logic can be tested. Whereas `build-script-impl` needed to be run in order to verify its behavior, the logic extracted out of it into `buildcli` can be tested in isolation.
- The logic can be split up into several files without polluting the `utils` directory, which now contains many different files that are unrelated to `build-script`.

/cc @gottesmm and @jrose-apple, since they've commented on [SR-237](https://bugs.swift.org/browse/SR-237) in the past.

gribozavr commented on utils/buildcli/buildcli/clang.py:
```diff
@@ -0,0 +1,85 @@
+def _freebsd_release_date():
+    """
+    Return the release date for FreeBSD operating system on this host.
+    If the release date cannot be ascertained, return None.
```
I think it will return `0` instead, or sometimes throw an error, if `sysctl` returns a non-integer.

I'd just halt the execution at that point.  There is no reasonable way a caller could recover from an error in this function.  They either need the version and they call this function, or they don't, and they shouldn't be calling it.

gribozavr commented on utils/buildcli/buildcli/clang.py:
```diff
@@ -0,0 +1,85 @@
+from __future__ import absolute_import
```
Please add Swift license headers to all new files with code (you can copy from `build-script`).

modocache replied:
Haha, I just had to be difficult about this, so I sent up #762. :sweat_smile:

For now, as you suggest I'll copy the header from `utils/build-script`.

gribozavr commented on utils/buildcli/buildcli/xcrun.py:
```diff
@@ -0,0 +1,18 @@
+        # `xcrun --find` prints to stderr when it fails to find the
+        # given tool. We swallow that output with a pipe.
+        out = subprocess.check_output(
+            ['xcrun', '--sdk', 'macosx', '--toolchain', 'default',
```
Please don't hardcode the toolchain.  `build-script-impl` accepts a parameter called `darwin-xcrun-toolchain` that should be honored for every call to `xcrun`.

modocache replied:
Ah, good catch. In order to respect the toolchain parameter, I may have to move the other `xcrun --find` out of `build-script-impl` as well, and change `build-script` such that it can take the toolchain as an argument.

Thanks for pointing this out!

gribozavr replied:
Or, for staging, you could make `build-script` interpret that parameter.  Whatever is easier.

gribozavr commented on utils/buildcli/README.md:
```diff
@@ -0,0 +1,10 @@
+# builcli
```
`buildcli` sounds like a cryptic name to me.  Why not use `SwiftBuildSupport`?

modocache replied:
Yeah, I agree that it's a little cryptic. I was wondering what to name it, and consulted [PEP-0008](https://www.python.org/dev/peps/pep-0008/):

> Modules should have short, all-lowercase names. Underscores can be used in the module name if it improves readability. Python packages should also have short, all-lowercase names, although the use of underscores is discouraged.

`buildcli` meets the "short, all-lowercase" criteria, but isn't very readable. How about `swift_build_support`?

gribozavr replied:
SGTM!

gribozavr commented on utils/buildcli/README.md:
````diff
@@ -0,0 +1,10 @@
+You may run unit tests for `buildcli` from the command line:
+
+```sh
+apple/swift $ python -m unittest discover -s utils/buildcli
````
It would be great if you could turn it into a test in the `test` directory.

modocache replied:
Ooh, great idea!! Will do. I'll also considering doing this for the tests I added to `utils/cmpcodesize`. Thanks for the tip!!

gribozavr replied:
`gyb` also has a self-test mode.

modocache replied:
Hmm, `utils/gyb --test` hangs indefinitely for me... :warning:

gribozavr replied:
Yes, that looks broken indeed.  https://bugs.swift.org/browse/SR-364

gribozavr commented on utils/buildcli/buildcli/clang.py:
```diff
@@ -0,0 +1,85 @@
+            # available.
+            return _first_clang()
+    else:
+        return _first_clang([''] + _CLANG_SUFFIXES)
```
The suffixes on Linux and FreeBSD are not the same.

I think it would be best if `_first_clang()` didn't have a defaulted parameter because of that.

modocache replied:
Yup, gotcha. I'll rethink this based on this and your comments above. Instead of reusing a list of suffixes, I'll specify them separately for each function call. I'll also remove the default parameter from `_first_clang()`--using default array/dict parameters [opens up a lot of room for gotchas](http://docs.python-guide.org/en/latest/writing/gotchas/#mutable-default-arguments), so it's for the best.

gribozavr commented on utils/buildcli/buildcli/clang.py:
```diff
@@ -0,0 +1,85 @@
+# host_clang() prefers versions of Clang in this order.
+# Note that this array must be updated when new, major versions of Clang
+# are released.
+_CLANG_SUFFIXES = ['38', '37', '36', '35']
```
This should be separate for Linux and FreeBSD.

Or maybe it should include both kinds of suffixes.

gribozavr commented on utils/buildcli/buildcli/clang.py:
```diff
@@ -0,0 +1,85 @@
+        if cc_path and cxx_path:
+            return CompilerExecutable(cc=cc_path, cxx=cxx_path)
+
+    return CompilerExecutable(cc=None, cxx=None)
```
Instead of introducing an invalid state in the `CompilerExecutable` object, why not just say that it is always valid, and signal errors from this function by returning `None`?

modocache replied:
Will do!

gribozavr commented:
@modocache Awesome improvements!

modocache commented:
@gribozavr Addressed your comments, except for the one about adding the unit tests to `validation-test`. That change may end up being a little more involved, so I think it's best to send it in another PR.

gribozavr commented on utils/swift_build_support/README.md:
```diff
@@ -0,0 +1,10 @@
+# builcli
```
Stale module name.

modocache replied:
Good catch!

gribozavr commented on utils/build-script:
```diff
@@ -497,6 +504,10 @@ the number of parallel build jobs to use""",
+    parser.add_argument("--darwin-xcrun-toolchain",
+                        help="the name of the toolchain to use on Darwin",
+                        default="default")
+
```
Sorry, but this is not a backwards-compatible change.  Previously, `--darwin-xcrun-toolchain ` was an argument that was passed after `--`:

```
$ ./utils/build-script -RT -- --darwin-xcrun-toolchain=foobar --other --flags
```

In order for the argument parsing to catch it now, it needs to be passed before `--`:

```
$ ./utils/build-script -RT --darwin-xcrun-toolchain=foobar --  --other --flags
```

I'm totally on board with breaking backward compatibility *once* when the refactoring is done, but not continuously throughout the process.

Thus, for staging, it seems like what we need to do is to preprocess the command line like this:

1. Filter out `--darwin-xcrun-toolchain.*` arguments from the command line into a separate list.

2. Append that list right before `--`.

3. Pass the new list to `parser.parse_args()`.

I think this pattern will appear over and over again as you migrate handling of other command-line arguments to Python, so it would make sense to make it a reusable function.

gribozavr commented:
@modocache Thank you!  Unfortunately, one more issue with argument handling came up in my testing...

modocache commented:
@gribozavr Good point, I'll update `utils/build-script` as per your suggestions.

I also realized I forgot to add the code headers you mentioned! I added the ones from #762, since it looks like that was just merged. Thanks!

modocache commented:
@gribozavr I've added some logic to the argument parser that enables the backwards-compatibility you described. I wanted to make sure we could test the new behavior in isolation, which meant moving the argument parser itself. Hopefully the commits aren't so big that they're unreviewable... :worried: 

Let me know if you have any additional thoughts--and thanks for all the reviews!

modocache commented:
A proposal to add Python unit tests to the validation tests is up at #778.

gribozavr commented:
@modocache Thanks!  I think your patch will work, but... Sorry for sounding picky, but I think we can use a simpler approach.  Subclassing of `argparse.ArgumentParser` seems strange to me (I don't feel like it was meant for subclassing), and `_move_impl_args` mutating the parsed arguments data structure after the fact seems error-prone (for example, because the information about the key in is duplicated).

What do you think about the following approach?  Add a function in the module (so that it is testable) that accepts a command line and a list of options to migrate, and returns the rewritten command line, which is then passed to `parse_args`.  Rewriting algorithm would be:

1. Split the argument list on `--`, to get two lists, one is the `build-script` arguments, another is the `build-script-impl` arguments.

2. Iterate over the `build-script-impl` argument list, and check if an option is marked for migration.  If it is, append it to the `build-script` argument list verbatim.

3. Concatenate two lists with `--` in between.

modocache commented:
@gribozavr Updated as per your suggestions! :+1: 

I think I'd still like to move the argument parser out into its own module someday, though. We could probably discuss this on the mailing lists, in the comments on [SR-237](https://bugs.swift.org/browse/SR-237), or in another pull request. My reasoning is that it's pretty cool to be able to call `parser.parse_args()` from within a unit test to verify what values are set. Even with the `test_migration.py` tests from this pull request, I still had to manually call `build-script`--the unit tests showed the `migrate_impl_args()` function was working, but I couldn't be sure the argument parsing was the same as before.

Anyway, thanks for all the dedicated reviews!

gribozavr commented:
@modocache Yes, that would be definitely useful -- I was thinking about a feature in build-script that would validate all presets, for example.  I wouldn't mind if you move it to a module in this PR, but it is the subclassing that looks strange -- I could imagine a function that accepts a list of command line arguments and returns a parsed representation as the API that the module should expose for this purpose.

modocache commented:
Ah, gotcha. Well, definitely expect more PRs on this topic, then! :wink: 

I feel like this one is in a pretty good state, though. I'm eager to merge it so I can add its tests to the validation tests from #778.

gribozavr commented on utils/swift_build_support/swift_build_support/clang.py:
```diff
@@ -0,0 +1,90 @@
+    """
+    for suffix in suffixes:
+        cc_path = which('clang{}'.format(suffix))
+        cxx_path = which('clang{}'.format(suffix))
```
This should say `clang++` -- I'm getting strange linker errors because CMake is trying to link C++ programs with the C compiler.

modocache replied:
Oops! Good catch, thanks--this was a typo. :blush:

gribozavr commented:
@modocache Thank you!  This patch is much simpler and I like it.  There's one issue though, the C++ compiler name is not correct, see above.

modocache commented:
@gribozavr Updated! Thanks.

modocache commented:
Also went ahead and added the tests to `validation-test`. :+1:

gribozavr commented:
@modocache Thank you!  Your PR passes all tests I did.  Merging.
